x86: intel64: implement runtime page tables

tyrantlizard77 opened the pull request:
What's changed:

- locore area split into text, rodata, data areas for proper MMU permissions
- necessary linker symbols added to intel64 linker script to compute MMU boot regions, with page alignment if runtime page tables are enabled
- MMU_BOOT_REGION() now uses Z_STRUCT_SECTION_ITERABLE()
- Enable MMU_BOOT_REGION() section for intel64
- Bring in CONFIG_X86_MMU and some related defines to common X86 Kconfig
- Don't set ARCH_HAS_USERSPACE or ARCH_HAS_STACK_PROTECTION for intel64 yet when X86_MMU is enabled, we're not there yet
- Uplevel x86_mmu.c to common code 
- Add logic for handling PML4 table and generating PDPTs
- move z_x86_paging_init() to common kernel_arch_func.h
- Uplevel inclusion of mmustructs.h to common x86 arch.h,
  both need it for memory domain defines
- Enable runtime page tables on boot CPU and SMP CPUs as they are brought up

Currently investigating some sporadic fatal errors that seem to be related to when CPU 1 brings up a thread, but do not have a reliable way to reproduce or ideas on the root cause yet. For now, X86_MMU can only be enabled if MP_NR_CPUS=1.

Fixes: #17832

Commit:
x86: generate runtime 64-bit page tables

- Bring in CONFIG_X86_MMU and some related defines to
  common X86 Kconfig
- Don't set ARCH_HAS_USERSPACE for intel64 yet when
  X86_MMU is enabled
- Uplevel x86_mmu.c to common code
- Add logic for handling PML4 table and generating PDPTs
- move z_x86_paging_init() to common kernel_arch_func.h
- Uplevel inclusion of mmustructs.h to common x86 arch.h,
  both need it for memory domain defines

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: intel64: set page tables for every CPU

The page tables to use are now stored in the cpuboot struct.
For the first CPU, we set to the flat page tables, and then
update later in z_x86_prep_c() once the runtime tables have
been generated.

For other CPUs, by the time we get to z_arch_start_cpu()
the runtime tables are ready do go, and so we just install
them directly.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
locore: organize data by type

Program text, rodata, and data need different MMU
permissions. Split out rodata and data from the program
text, updating the linker script appropriately.

Region size symbols added to the linker script, so these
can later be used with MMU_BOOT_REGION().

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: intel64: set up memory sections

Some linker symbols indicating the bounds of various memory
areas needed for setting up boot MMU regions are now in
place.

MMU_BOOT_REGION macros added to align the bounds of these
sections if we're using runtime page tables.

_image_rodata_start moved to the proper place, it was not
accounting for the rodata sections specified in
linker/common_rom.ld.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: mmustructs: use Z_STRUCT_SECTION_ITERABLE()

This does the right thing for arches with 8-byte words.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: intel64: enable MMU region list

Same snippet as in our 32-bit linker file. Creates
an iterable section with all the MMU_BOOT_REGION() info.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
Passing CI results here are misleading -- a typical local sanitycheck run will have a dozen sporadic failures that simply get fixed when CI retries them.

gnuless commented on arch/x86/core/intel64/locore.S:
```diff
@@ -622,14 +455,101 @@ IRQ(232); IRQ(233); IRQ(234); IRQ(235); IRQ(236); IRQ(237); IRQ(238); IRQ(239)
 
-pml4:
+.globl z_x86_flat_ptables
+z_x86_flat_ptables:
```
You probably want to remove the G bit on the entries here.  Unless you explicitly invalidate these TLBs they will likely hang around on the BSP forever.  

> pdp:	.long 0x00000183		/* 0x183 = G, 1GB, R/W, P */

andrewboie replied:
done.

gnuless commented:
As I mentioned in email, I suggest deferring the SMP issues to another day.  Better to prohibit the selection of SMP and memory protection in the same build for now.

Commit:
intel64: don't set global flag in ptables

May cause issues when runtime page tables are installed.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
REVERTME: x86: disable X86_MMU if MP_NUM_CPUS != 1

We are working through some SMP issues with this enabled,
prevent this for now.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
> As I mentioned in email, I suggest deferring the SMP issues to another day. Better to prohibit the selection of SMP and memory protection in the same build for now.

Done.
I added a new "qemu_x86_long_up" target which has a single CPU and X86_MMU enabled, so that this mechanism is still tested.

andrewboie commented:
> I added a new "qemu_x86_long_up" target which has a single CPU and X86_MMU enabled, so that this mechanism is still tested.

Dropped this -- it's extra CI load for something that is transient.
The restriction on X86_MMU depending on MP_NUM_CPUS=1 is in place.
